internal/validate/validate.go

```diff
@@ -11,19 +11,51 @@ import (
 // Validate that the values in the Request struct are valid according to the
 // validation rules defined on the struct.
 // If validation fails the error will be of type Error.
+//
+// Validate automatically traverses the fields on the struct. If any of the
+// fields are of a type that implement Request, the validation rules of that
+// field will be used as well.
 func Validate(req Request) error {
+	reqV := reflect.Indirect(reflect.ValueOf(req))
+	err := validateStruct(reqV)
+	if len(err) > 0 {
+		return err
+	}
+	return nil
+}
+
+func validateStruct(v reflect.Value) Error {
 	err := make(Error)
 
-	for _, rule := range req.ValidationRules() {
-		if failure := rule.Validate(); failure != nil {
-			err[failure.Name] = append(err[failure.Name], failure.Problems...)
+	req, ok := v.Interface().(Request)
+	if ok {
+		for _, rule := range req.ValidationRules() {
+			if failure := rule.Validate(); failure != nil {
+				err[failure.Name] = append(err[failure.Name], failure.Problems...)
+			}
 		}
 	}
 
-	if len(err) > 0 {
-		return err
+	if v.Kind() == reflect.Struct {
+		for i := 0; i < v.NumField(); i++ {
+			f := v.Field(i)
+			if v.Type().Field(i).Anonymous {
```

Comment:
What does this anonymous check do?

Reply:
`Anonymous` means the field is an embedded structs: https://pkg.go.dev/reflect#StructField

I can add a comment to say that

Reply:
Thanks for the explanation